### PR TITLE
Async support to prevent deadlocks

### LINKED ISSUE #40: Implement async eager methods

### PR DESCRIPTION
**Pull request by Mafii**
Instead of `.ToList()`, we can now use `await .ToListAsnyc()`

Resolves #40

### CODE REVIEW

**Comment by Mafii** [comment]
@bash does this resolve #40?

**Comment by tautropfli** [comment]
Yes, this resolves #40

File: LinqToRest/Async/IAsyncQueryProvider.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    /// <summary>
+    ///     <para>
+    ///         Defines method to execute queries asynchronously that are described by an IQueryable object.
+    ///     </para>
+    ///     <para>
+    ///         The service lifetime is <see cref="ServiceLifetime.Scoped" />. This means that each
+    ///         <see cref="DbContext" /> instance will use its own instance of this service.
+    ///         The implementation may depend on other services registered with any lifetime.
+    ///         The implementation does not need to be thread-safe.
+    ///     </para>
```
Comment (tautropfli) [inline LinqToRest/Async/IAsyncQueryProvider.cs]: Remove this comment as it only applies to ASP.NET

File: LinqToRest/Async/Infrastructure/FakeAsyncEnumerable.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            foreach (var item in _source)
+            {
+                yield return item;
+            }
+
+            await Task.CompletedTask;
+        }
+
+        public IEnumerator<T> GetEnumerator()
```
Comment (tautropfli) [inline LinqToRest/Async/Infrastructure/FakeAsyncEnumerable.cs]: Use an expression body

File: LinqToRest/Async/Infrastructure/QueryableMethods.cs [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+using System.Linq.Expressions;
+using System.Reflection;
+using JetBrains.Annotations;
+
+namespace Messerli.LinqToRest.Async.Infrastructure
+{
+    /// <summary>
+    ///     A class that provides reflection metadata for translatable LINQ methods.
+    /// </summary>
+    public static class QueryableMethods
```
Comment (tautropfli) [inline LinqToRest/Async/Infrastructure/QueryableMethods.cs]: Could this be internal?
```suggestion
    internal static class QueryableMethods
```

File: LinqToRest/Async/Resources/AbstractionsStrings.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+﻿// from https://github.com/dotnet/efcore/blob/master/src/EFCore.Abstractions/Properties/AbstractionsStrings.Designer.cs
```
Comment (tautropfli) [inline LinqToRest/Async/Resources/AbstractionsStrings.cs:1]: Check if this is generated automatically on build and add to .gitignore

File: LinqToRest/Async/Resources/CoreStrings.cs [hunk 5]
```diff
@@ -0,0 +1,2743 @@
+﻿// from https://github.com/dotnet/efcore/blob/master/src/EFCore/Properties/CoreStrings.Designer.cs
```
Comment (tautropfli) [inline LinqToRest/Async/Resources/CoreStrings.cs]: Check if this is generated automatically on build and add to .gitignore

File: LinqToRest/LinqToRest.csproj [hunk 6]
```diff
@@ -27,5 +27,11 @@
         <PackageReference Include="Messerli.ServerCommunication" />
         <PackageReference Include="Messerli.QueryProvider" />
         <PackageReference Include="Newtonsoft.Json" />
+        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
+    </ItemGroup>
+    <ItemGroup>
+      <EmbeddedResource Update="Async\Resources\CoreStrings.resx">
+        <DependentUpon>CoreStrings.cs</DependentUpon>
+      </EmbeddedResource>
```
Comment (tautropfli) [inline LinqToRest/LinqToRest.csproj]: Is this here intentionally?

File: LinqToRest/ProjectionReader.cs [hunk 7]
```diff
@@ -1,26 +1,26 @@
-using System;
```
Comment (tautropfli) [inline LinqToRest/ProjectionReader.cs]: Why is this file modified?

File: LinqToRest/QueryProvider.cs [hunk 8]
```diff
@@ -34,15 +39,34 @@ public override object Execute(Expression expression)
             return Activator.CreateInstance(typeof(ProjectionReader<>).MakeGenericType(elementType), _resourceRetriever, uri);
         }
 
+        public TResult ExecuteAsync<TResult>(Expression expression, CancellationToken cancellationToken = default)
+        {
+            // Validate Generic Type is a Task (might need to be adjusted if we need to support ValueTasks as well)
+            if (!(typeof(TResult).GetGenericTypeDefinition() == typeof(Task<>)))
```
Comment (tautropfli) [inline LinqToRest/QueryProvider.cs]: Extract predicate into a function

File: LinqToRest/QueryProvider.cs [hunk 9]
```diff
@@ -34,15 +39,34 @@ public override object Execute(Expression expression)
...
+            {
+                throw new ArgumentException("Type is expected to be a generic Task type.");
```
Comment (tautropfli) [inline LinqToRest/QueryProvider.cs]: Add the actual type to the exception message

File: Packages.props [hunk 10]
```diff
@@ -1,22 +1,23 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (tautropfli) [inline Packages.props]: Convert this to unix newlines

File: LinqToRest/QueryProvider.cs [hunk 11]
```diff
@@ -34,15 +39,34 @@ public override object Execute(Expression expression)
...
+
+            return (TResult) methodToExecute.Invoke(_resourceRetriever, new object[] {uri, cancellationToken});
+        }
+
         private TranslateResult Translate(Expression expression)
         {
             expression = Evaluator.PartialEval(expression);
             var queryBinder = _queryBinderFactory();
-            var proj = (ProjectionExpression)queryBinder.Bind(expression);
+            var proj = (ProjectionExpression) queryBinder.Bind(expression);
```
Comment (tautropfli) [inline LinqToRest/QueryProvider.cs]: Remove the space between type and value:
```suggestion
            var proj = (ProjectionExpression)queryBinder.Bind(expression);
```

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: LinqToRest/Async/IAsyncQueryProvider.cs (hunk 1 above)
Comment (Mafii) [reply LinqToRest/Async/IAsyncQueryProvider.cs]: Fixed in bf28bde6d322597e5f66797777608ad663695518

File: LinqToRest/Async/Infrastructure/FakeAsyncEnumerable.cs (hunk 2 above)
Comment (Mafii) [reply LinqToRest/Async/Infrastructure/FakeAsyncEnumerable.cs]: Fixed in f86d9cc2f36b5e9b66da30226c82ea0a5ada1b24

File: LinqToRest/Async/Infrastructure/QueryableMethods.cs (hunk 3 above)
Comment (Mafii) [reply LinqToRest/Async/Infrastructure/QueryableMethods.cs]: Fixed in af513b81e206a3058dd49bfb5fa2723dec122588

File: LinqToRest/Async/Resources/AbstractionsStrings.cs (hunk 4 above)
Comment (Mafii) [reply LinqToRest/Async/Resources/AbstractionsStrings.cs:1]: Decision to not do this, as we would need to add T4 generation to the build, thus adding unessecary complexity, since we don't need to update resources that are copied from EF Core

File: LinqToRest/ProjectionReader.cs (hunk 7 above)
Comment (Mafii) [reply LinqToRest/ProjectionReader.cs]: Fixed in 379eb22827cd2968cbad74f6b4209d51eeb05254

**Review by tautropfli — APPROVED** [review]
